dimod/constrained.py

```diff
@@ -1959,6 +1959,35 @@ def var_encoder(v):
             return f'{vartype_name[self.vartype(v)]}({v!r})'
 
         sio = StringIO()
+
+        def render_limited_number(iterable, render_element):
+            limited = False
+            last = None
+
+            for k, x in enumerate(iterable):
+                assert x is not None
+
+                if last is not None:
+                    if k < 4:
```

Comment:
IMO this number could be higher. More like 10 or even 20.

Reply:
👍 
Maybe the number of head and tail items should be symmetric also?  That would be consistent with both `numpy` and `pandas` (and shouldn't be too hard to implement):

```
>>> import pandas as pd
>>> pd.DataFrame({str(x): [y*x for y in range(100)] for x in range(100)})
    0   1    2    3    4    5  ...    94    95    96    97    98    99
0   0   0    0    0    0    0  ...     0     0     0     0     0     0
1   0   1    2    3    4    5  ...    94    95    96    97    98    99
2   0   2    4    6    8   10  ...   188   190   192   194   196   198
3   0   3    6    9   12   15  ...   282   285   288   291   294   297
4   0   4    8   12   16   20  ...   376   380   384   388   392   396
.. ..  ..  ...  ...  ...  ...  ...   ...   ...   ...   ...   ...   ...
95  0  95  190  285  380  475  ...  8930  9025  9120  9215  9310  9405
96  0  96  192  288  384  480  ...  9024  9120  9216  9312  9408  9504
97  0  97  194  291  388  485  ...  9118  9215  9312  9409  9506  9603
98  0  98  196  294  392  490  ...  9212  9310  9408  9506  9604  9702
99  0  99  198  297  396  495  ...  9306  9405  9504  9603  9702  9801

[100 rows x 100 columns]
>>> import numpy as np
>>> np.array([[y*x for y in range(100)] for x in range(100)])
array([[   0,    0,    0, ...,    0,    0,    0],
       [   0,    1,    2, ...,   97,   98,   99],
       [   0,    2,    4, ...,  194,  196,  198],
       ...,
       [   0,   97,  194, ..., 9409, 9506, 9603],
       [   0,   98,  196, ..., 9506, 9604, 9702],
       [   0,   99,  198, ..., 9603, 9702, 9801]])
```

BTW, FWIW, I did spend a few minutes looking for the functions that in both `numpy` and `pandas` that produce those outputs, so that we might reuse them.  Came up empty in both cases... I am guessing it's done in C code, so we can't reuse.

Reply:
I believe it's [here](https://github.com/numpy/numpy/blob/main/numpy/core/arrayprint.py). We used some of it for [sample set printing](https://github.com/dwavesystems/dimod/blob/main/dimod/serialization/format.py).

But yes, I like the centered format.

Reply:
OK, I didn't reuse any of that, but did get the centered format and introduced a private class attribute to control the number of display items, with default value = 10.